### PR TITLE
Product subscriptions: Free trial section

### PR DESCRIPTION
**Pull request by 0nko**
Implements #10148, which adds a new Subscription free trial to the product detail and variation detail screen.

| Product detail | Subscription free trial |
| --- | --- |
| image | image |

**To test:**

Simple subscription product
1. Make sure you have the Subscriptions plugin installed
2. Prepare a simple subscription product
3. Start the app
4. Go to Products -> Tap on the simple subscription product
5. Notice there's a new Subscription free trial section with the correct data shown
6. Tap on the free trial property
7. Notice a detail free trial screen is shown with the correct data
8. Try changing the values
9. Tap back
10. Notice the free trial section information was update

Variable subscription product
1. Make sure you have the Subscriptions plugin installed
2. Prepare a variable subscription product (make sure you generate some variations)
3. Start the app
4. Go to Products -> Tap on the variable subscription product
5. Tap on Variations -> Select a variation
6. Notice there's a new Subscription free trial section with the correct data shown
7. Tap on the free trial property
8. Notice a detail free trial screen is shown with the correct data
9. Try changing the values
10. Tap back
11. Notice the free trial section information was update

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt [hunk 1]
```diff
@@ -80,28 +87,45 @@ class ProductSubscriptionExpirationFragment : BaseProductFragment() {
         items: List<String>,
         currentValue: String
     ) {
-        Box(
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt:83]: This is unrelated to the free trial but adds the grey background to the subscription screen for consistency.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 2]
```diff
@@ -767,49 +789,49 @@ class ProductDetailViewModel @Inject constructor(
         var eventName: AnalyticsEvent? = null
         var hasChanges = false
         when (event) {
-            is ProductExitEvent.ExitSettings -> {
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt:770]: np, I think these changes were an accidental refactoring by the IDE, scoping the events to the parent class `ProductExitEvent` is easier to read IMO, so I suggest reverting it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigator.kt [hunk 3]
```diff
@@ -314,8 +322,8 @@ class ProductNavigator @Inject constructor() {
                     directions = directions,
                     navOptions =
                     if (target.source == STORE_ONBOARDING)
-                        NavOptions.Builder()
-                            .setPopUpTo(R.id.dashboard, false)
+                        Builder()
+                            .setPopUpTo(id.dashboard, false)
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigator.kt]: Same as above, I think this was an accidental refactoring by the IDE, `NavOptions.Builder` is clearer than just `Builder`.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigator.kt [hunk 4]
```diff
@@ -409,36 +417,44 @@ class ProductNavigator @Inject constructor() {
                 )
             }
 
-            is ProductNavigationTarget.ViewProductSubscription -> {
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigator.kt]: Same as above 🙂

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt (hunk 1 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt:83]: Thanks for this @0nko, but I have a minor remark the usage of the `Divider` below, in our current screens, it's not a divider, but a shadow of the Card that holds the content, we can follow the same if we want, WDYT? It's a minor point.

Just for information, I experienced with this layout:
```kotlin
Box(Modifier.fillMaxSize()) {
            Card(
                shape = RectangleShape,
                modifier = Modifier
                    .fillMaxWidth()
            ) {
                Row(
                    modifier = Modifier
                        .padding(dimensionResource(id = dimen.major_100)),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    OutlinedTextField(
                        value = if (length == 0) "" else length.toString(),
                        onValueChange = { input ->
                            val trialLength = if (input.isNotEmpty()) {
                                input.filter { it.isDigit() }.toInt()
                            } else {
                                0
                            }
                            onTrialLengthUpdated(trialLength)
                        },
                        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                        modifier = Modifier
                            .weight(.3f)
                            .padding(end = dimensionResource(id = dimen.major_100))
                    )
                    WcExposedDropDown(
                        items = items.toTypedArray(),
                        onSelected = { onTrialPeriodUpdated(it) },
                        currentValue = period,
                        mapper = { it.getPeriodString(resourceProvider, length) },
                        modifier = Modifier
                            .weight(.7f)
                            .background(colorResource(id = color.color_surface))
                            .padding(start = dimensionResource(id = dimen.major_100))
                    )
                }
            }
        }
```

and it gives the following:

**Review by hichamboushaba — COMMENTED** [review]
Nice work @0nko, this works well.

I left some comments, mostly minor about IDE auto refactoring that messed up some imports.
I'm not approving yet as I see we need to implement some client side validation p1700049601331839-slack-C03L1NF1EA3, if you are planning on adding on a separate PR, please let me know.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt (hunk 1 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt:83]: Thanks for the suggestion! You're right about this. I copied what I saw it some layout XML but this is the right approach 👍.

**Comment by 0nko** [comment]
Thank you for the review @hichamboushaba! I've addressed the feedback and added range validation. Ready for another round.

<image src="https://github.com/woocommerce/woocommerce-android/assets/1522856/c7ef2825-5274-424e-b554-cdfdc867d8d1" width="300" />

**Review by hichamboushaba — APPROVED** [review]
Nice job @0nko, :shipit: 

Since you are AFK today, I'll fix conflicts then merge the PR

**Comment by 0nko** [comment]
Thank you @hichamboushaba!

**Pull request merged**